Text completer

Lastrellik opened the pull request:
After learning about the QCompleter tool from @ecdsa, I came up with a CompletionTextEdit class that has been utilized in both paytoedit.py and seed_dialog.py. This feature allows the user to optionally select from a list of words when restoring from seed in order to make the process a little easier and less error prone. 

Before merging this pull request, if someone could make sure that the send functionality in the main window still auto populates, that would be great :) I know that it doesn't crash, but I didn't have anyone in my address book that I could test it with :) I appreciate your patience.

bauerj commented on gui/qt/paytoedit.py:
```diff
@@ -23,14 +23,12 @@
-from PyQt5.QtWidgets import QCompleter, QPlainTextEdit
-from .qrtextedit import ScanQRTextEdit
-
+from .completion_text_edit import CompletionTextEdit
```
Please move this import down to the other imports from Electrum.

bauerj commented on gui/qt/seed_dialog.py:
```diff
@@ -131,6 +132,13 @@ def __init__(self, seed=None, title=None, icon=True, msg=None, options=None, is_
         self.addWidget(self.seed_warning)
 
+    def initialize_completer(self):
+        english_path = os.path.join(os.path.dirname(__file__), "../../lib/wordlist/english.txt")
```
Why don't you use a `Mnemonic` object?

    Mnemonic('en').wordlist

Lastrellik commented:
@bauerj that last commit should take care of the suggestions in your review. I have to be done for the day, but I can work on disabling the suggestions tomorrow, if you'd like.

Lastrellik commented:
I forgot to mention, this is in reference to [#1116](https://github.com/spesmilo/electrum/issues/1116)

SomberNight commented on gui/qt/seed_dialog.py:
```diff
@@ -131,6 +133,13 @@ def __init__(self, seed=None, title=None, icon=True, msg=None, options=None, is_
+    def initialize_completer(self):
+        english_list = Mnemonic('en').wordlist
+        old_list = electrum.old_mnemonic.words
+        wordlist = english_list + list(set(old_list) - set(english_list))
```
maybe this is easier to read:
`wordlist = list(set(old_list) | set(english_list))`

SomberNight replied:
but if you are doing that to have the words of the new list appear before the words of the old, then disregard

Lastrellik replied:
It was more to concatenate both lists without having duplicates. Because we're not doing any sort of binary search or tree data structure, the order of the list doesn't matter.

SomberNight replied:
What does the order of the suggestions presented to the user depend on? Is that independent of the order in this list here?

SomberNight commented on gui/qt/completion_text_edit.py:
```diff
@@ -0,0 +1,108 @@
+                return True
+        if e.key() in [Qt.Key_Tab]:
+            return True
+        if e.key() in [Qt.Key_Down, Qt.Key_Up] and not self.is_multiline():
```
`is_multiline` was defined in `PayToEdit` but not here

SomberNight commented on gui/qt/paytoedit.py:
```diff
@@ -177,9 +177,6 @@ def get_outputs(self, is_max):
     def lines(self):
         return self.toPlainText().split('\n')
 
-    def is_multiline(self):
```
this method is used in this class, around line 206 :P

Lastrellik replied:
I saw that. My thought is that it does extend CompletionTextEdit, so that functionality should still be available. Unless I need to use ```parent.is_multiline()``` instead of ```self.is_multiline()``` on line 206

SomberNight replied:
It should be fine; sorry, I was not paying enough attention. See EDIT above.

Lastrellik commented:
@SomberNight thanks for your feedback! I went ahead and made those changes in my last commit.

Lastrellik commented:
I should have been paying closer attention :) The is_multiline only returns a value that is more than one if the text contains ```\n```. This only applies to paytoedit.py. So in this last commit I moved ```is_multiline()``` back to paytoedit.py and I also added a feature where the suggestions are disabled if the user enters a word that is not in the wordlist, as requested by @bauerj

SomberNight commented on gui/qt/seed_dialog.py:
```diff
@@ -160,6 +160,13 @@ def on_edit(self):
+                self.seed_e.disable_suggestions()
+                break
+            else:
+                self.seed_e.enable_suggestions()
```
it might make more sense for the `else` to be one indentation layer "outer"; so that it accompanies the `for` loop instead of the `if`

Lastrellik commented:
@SomberNight thanks for the suggestion :) I went ahead and made that change with that last commit. I appreciate everyone's patience while I get familiar with Electrum and its code base :)

SomberNight commented:
If the current partial word is a valid word (present in the dict) and you press enter, it will duplicate the word.
Can you make the suggestions disappear and not do anything in that case?

<details>
 <summary>pics</summary>

![seed_word_suggestions2](https://user-images.githubusercontent.com/29142493/37123007-f198861c-2262-11e8-97d5-8e5315f2df67.PNG)
![seed_word_suggestions3](https://user-images.githubusercontent.com/29142493/37123049-1b58f400-2263-11e8-8e8b-74c4008140d5.PNG)
</details>

Also, by the way, I think for the seed word case, it might make sense to also add a whitespace when the user auto-completes a word, so they can start entering the next word right away. (whitespaces are normalized out from seeds so don't worry about not adding a whitespace after the last word or stuff like that)

In fact, it might be easier to solve the first thing while solving the second:
maybe you can append a whitespace after every word in the dictionary or something like that;
or do it some other way :P

Lastrellik commented:
@SomberNight again, thank you for your suggestions! I think we're all good now. What do you think?

SomberNight commented:
Great; thanks!
@ecdsa I think this can get merged
